### DISCUSSION TITLE
InlayHints support for gopls

### DISCUSSION THREAD

**Discussion by Gavinok** [joaotavora/eglot#1369] [Q&A] [answered]
I was wondering if anyone has gotten inlay hints working with gopls?

 I know the server supports it but can't get them to display after enabling them via `eglot-inlay-hints-mode` I was able to set it up with typescript mode but cant figure out how to translate the settings for gopls to work with eglot

https://github.com/golang/tools/blob/master/gopls/doc/settings.md#inlayhint

Shows how the settings should work am just unable to enable it myself.

**Comment by Gavinok** [comment]
Managed to get this working in the end. Hope this can be of help to future gophers

```elisp
(use-package eglot-gopls-x :no-require t
  :after (:all eglot go-ts-mode)
  :config
  ;; Ensure gopls will use typehints (IDK why I needed to do this manually)
  (setq-default eglot-workspace-configuration
                '(:gopls (:hints (
                                  :assignVariableTypes t
                                  :compositeLiteralFields t
                                  :compositeLiteralTypes t
                                  :constantValues t
                                  :functionTypeParameters t
                                  :parameterNames t
                                  :rangeVariableTypes t
                                  ))))

  (cl-defmethod eglot-client-capabilities :around (server)
    "Allow dynamic registration of inlay hints since it's the only way
to get gopls to give us any inlay hints"
    (let ((base (cl-call-next-method)))
      (when (cl-find "gopls" (process-command
                              (jsonrpc--process server))
                     :test #'string-match)
        (setf (cl-getf (cl-getf base :textDocument) :inlayHint)
              `(:dynamicRegistration t)))
      base))

  ;; Ensure inlay hints going for eglot
  (add-hook 'go-ts-mode-hook 'eglot-inlay-hints-mode))
  ```

**Reply by joaotavora** [reply]
>   "Allow dynamic registration of inlay hints since it's the only way to get gopls to give us any inlay hints"

It's bizarre that this work or needed since as far as I know, just stating you support dynamic registration is not enoough, you actually have to support it.  

Don't know what the `eglot-gopls-x` is, but I'd say you don't need it.

I get inlay hints working with gopls 0.15.1 with just this:
```
path/to/recent/emacs -Q /path/to/hello.go -f go-ts-mode \
--eval '(setq-default eglot-workspace-configuration (quote (:gopls (:hints (:parameterNames t)))))' \
-f eglot 
```

If `hello.go`is:
```
package main

import ("fmt")

func add(x int, y int) int { return x + y }

func main() {
        var num1, num2 int
        sum := add(num1, num2)
        fmt.Printf("The sum of %d and %d is %d\n", num1, num2, sum)
}
```

And I see inlay hints fine:
![image](https://github.com/joaotavora/eglot/assets/387011/e4e18027-d2d2-4c62-b34c-f5a7819e991b)

**Comment by joaotavora** [comment, marked as answer]
Sorry, @Gavinok , I unmarked you answer because I don't think it needs to be as complicated as you say.  

A much simpler way is to just set:

```
(setq-default eglot-workspace-configuration (quote (:gopls (:hints (:parameterNames t)))))
```

add more types of hints if you want.  See https://github.com/joaotavora/eglot/discussions/1369#discussioncomment-8779762 for a full `emacs -Q` recipe.  

Ideally, even this shouldn't be needed.  Gopls should enable this by default, or allow this via `initializationOptions` or command-line parameters, since this is not usually a per-project thing.

**Reply by Gavinok** [reply]
No problem maybe this was just a mistake on debugging this locally. The whole solution was a hack and I just figured I'd share what I had at the time.

**Reply by Gavinok** [reply]
thanks for your help

**Reply by joaotavora** [reply]
Thanks for understanding :+1: I've marked this the answer.

**Reply by Gavinok** [reply]
Just tested and this does work. Must have just over complicated things.
